File: uhrwerk-common/src/main/java/io/qimia/uhrwerk/common/model/TableModel.java
package io.qimia.uhrwerk.common.model;

import com.google.common.base.Objects;
import java.io.Serializable;
import java.util.Arrays;

public class TableModel extends StateModel implements Serializable {

  private static final long serialVersionUID = 8806220232347910731L;

  final String area;
  final String vertical;
  final String name;
  final String version;
  String className;
  int parallelism;
  int maxBulkSize;
  PartitionUnit partitionUnit;
  int partitionSize;
  boolean partitioned = false;

  DependencyModel[] dependencies;
  SourceModel[] sources;
  TargetModel[] targets;

  public static TableModelBuilder builder() {
    return new TableModelBuilder();
  }

  public TableModel(TableModelBuilder builder) {
    super(builder.deactivatedTs);
    this.id = builder.id;
    this.area = builder.area;
    this.vertical = builder.vertical;
    this.name = builder.name;
    this.version = builder.version;
    this.className = builder.className;
    this.parallelism = builder.parallelism;
    this.maxBulkSize = builder.maxBulkSize;
    this.partitionUnit = builder.partitionUnit;
    this.partitionSize = builder.partitionSize;
    this.partitioned = builder.partitioned;
    this.dependencies = builder.dependencies;
    this.sources = builder.sources;
    this.targets = builder.targets;
  }
  public String getArea() {
    return area;
  }

  public String getVertical() {
    return vertical;
  }

  public String getName() {
    return name;
  }

  public String getVersion() {
    return version;
  }

  public String getClassName() {
    return className;
  }

  public void setClassName(String className) {
    this.className = className;
  }

  public int getParallelism() {
    return parallelism;
  }

  public void setParallelism(int parallelism) {
    this.parallelism = parallelism;
  }

  public int getMaxBulkSize() {
    return maxBulkSize;
  }

  public void setMaxBulkSize(int maxBulkSize) {
    this.maxBulkSize = maxBulkSize;
  }

  public PartitionUnit getPartitionUnit() {
    return partitionUnit;
  }

  public void setPartitionUnit(PartitionUnit partitionUnit) {
    this.partitionUnit = partitionUnit;
  }

  public int getPartitionSize() {
    return partitionSize;
  }

  public void setPartitionSize(int partitionSize) {
    this.partitionSize = partitionSize;
  }

  public DependencyModel[] getDependencies() {
    return dependencies;
  }

  public void setDependencies(DependencyModel[] dependencies) {
    this.dependencies = dependencies;
  }

  public SourceModel[] getSources() {
    return sources;
  }

  public void setSources(SourceModel[] sources) {
    this.sources = sources;
  }

  public TargetModel[] getTargets() {
    return targets;
  }

  public void setTargets(TargetModel[] targets) {
    this.targets = targets;
  }

  public boolean isPartitioned() {
    return partitioned;
  }

  public void setPartitioned(boolean partitioned) {
    this.partitioned = partitioned;
  }

  @Override
  public boolean equals(Object o) {
    if (this == o) {
      return true;
    }
    if (!(o instanceof TableModel)) {
      return false;
    }
    if (!super.equals(o)) {
      return false;
    }
    TableModel table = (TableModel) o;
    return getParallelism() == table.getParallelism()
        && getMaxBulkSize() == table.getMaxBulkSize()
        && getPartitionSize() == table.getPartitionSize()
        && isPartitioned() == table.isPartitioned()
        && Objects.equal(getArea(), table.getArea())
        && Objects.equal(getVertical(), table.getVertical())
        && Objects.equal(getName(), table.getName())
        && Objects.equal(getVersion(), table.getVersion())
        && Objects.equal(getClassName(), table.getClassName())
        && getPartitionUnit() == table.getPartitionUnit()
        && Objects.equal(getDependencies(), table.getDependencies())
        && Objects.equal(getSources(), table.getSources())
        && Objects.equal(getTargets(), table.getTargets());
  }

  @Override
  public int hashCode() {
    return Objects.hashCode(
        super.hashCode(),
        getArea(),
        getVertical(),
        getName(),
        getVersion(),
        getClassName(),
        getParallelism(),
        getMaxBulkSize(),
        getPartitionUnit(),
        getPartitionSize(),
        isPartitioned(),
        getDependencies(),
        getSources(),
        getTargets());
  }

  @Override
  public String toString() {
    return "TableModel{" +
        "area='" + area + '\'' +
        ", vertical='" + vertical + '\'' +
        ", name='" + name + '\'' +
        ", version='" + version + '\'' +
        ", className='" + className + '\'' +
        ", parallelism=" + parallelism +
        ", maxBulkSize=" + maxBulkSize +
        ", partitionUnit=" + partitionUnit +
        ", partitionSize=" + partitionSize +
        ", partitioned=" + partitioned +
        ", dependencies=" + Arrays.toString(dependencies) +
        ", sources=" + Arrays.toString(sources) +
        ", targets=" + Arrays.toString(targets) +
        '}';
  }
}

File: uhrwerk-metastore/src/main/kotlin/io/qimia/uhrwerk/common/metastore/model/TableModel.kt
package io.qimia.uhrwerk.common.metastore.model

import io.qimia.uhrwerk.common.model.TargetModel
import java.sql.Timestamp
import java.time.LocalDateTime

data class TableModel(
    var id: Long? = null,
    var area: String? = null,
    var vertical: String? = null,
    var name: String? = null,
    var version: String? = null,
    var className: String? = null,
    var transformSqlQuery: String? = null,
    var functions: Array<FunctionCallModel>? = null,
    var partitionColumns: Array<String>? = null,
    var partitionMappings: Map<String, Any>? = null,
    var dynamicPartitioning: Boolean = false,
    var tableVariables: Array<String>? = null,
    var parallelism: Int? = null,
    var maxBulkSize: Int? = null,
    var partitionUnit: PartitionUnit? = null,
    var partitionSize: Int? = null,
    var partitioned: Boolean = false,
    var dependencies: Array<DependencyModel>? = null,
    var sources: Array<SourceModel2>? = null,
    var targets: Array<TargetModel>? = null,
    var hashKey: Long? = null,
    var description: String? = null,
    var deactivatedTs: LocalDateTime? = null,
    var createdTs: Timestamp? = null,
    var updatedTs: Timestamp? = null
) : BaseModel {
    override fun id(id: Long?) {
        this.id = id
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other !is TableModel) return false
        if (id != null && other.id != null) if (id != other.id) return false
        if (area != other.area) return false
        if (vertical != other.vertical) return false
        if (name != other.name) return false
        if (version != other.version) return false
        if (className != other.className) return false
        if (transformSqlQuery != other.transformSqlQuery) return false

        if (!functions.isNullOrEmpty()) {
            if (other.functions.isNullOrEmpty()) return false
            if (!functions.contentEquals(other.functions)) return false
        } else if (!other.functions.isNullOrEmpty()) return false

        if (!partitionColumns.isNullOrEmpty()) {
            if (other.partitionColumns.isNullOrEmpty()) return false
            if (!partitionColumns.contentEquals(other.partitionColumns)) return false
        } else if (!other.partitionColumns.isNullOrEmpty()) return false

        if (!tableVariables.isNullOrEmpty()) {
            if (other.tableVariables.isNullOrEmpty()) return false
            if (!tableVariables.contentEquals(other.tableVariables)) return false
        } else if (!other.tableVariables.isNullOrEmpty()) return false

        if (parallelism != other.parallelism) return false
        if (maxBulkSize != other.maxBulkSize) return false
        if (partitionUnit != other.partitionUnit) return false
        if (partitionSize != other.partitionSize) return false
        if (partitioned != other.partitioned) return false

        if (!dependencies.isNullOrEmpty()) {
            if (other.dependencies.isNullOrEmpty()) return false
            if (!dependencies.contentEquals(other.dependencies)) return false
        } else if (!other.dependencies.isNullOrEmpty()) return false

        if (!sources.isNullOrEmpty()) {
            if (other.sources.isNullOrEmpty()) return false
            if (!sources.contentEquals(other.sources)) return false
        } else if (!other.sources.isNullOrEmpty()) return false

        if (!targets.isNullOrEmpty()) {
            if (other.targets.isNullOrEmpty()) return false
            if (!targets.contentEquals(other.targets)) return false
        } else if (!other.targets.isNullOrEmpty()) return false

        if (hashKey != null && other.hashKey != null) if (hashKey != other.hashKey) return false
        if (description != other.description) return false
        if (deactivatedTs != other.deactivatedTs) return false
        if (createdTs != other.createdTs) return false
        if (updatedTs != other.updatedTs) return false

        return true
    }

    override fun hashCode(): Int {
        var result = id?.hashCode() ?: 0
        result = 31 * result + (area?.hashCode() ?: 0)
        result = 31 * result + (vertical?.hashCode() ?: 0)
        result = 31 * result + (name?.hashCode() ?: 0)
        result = 31 * result + (version?.hashCode() ?: 0)
        result = 31 * result + (className?.hashCode() ?: 0)
        result = 31 * result + (transformSqlQuery?.hashCode() ?: 0)
        result = 31 * result + (partitionColumns?.contentHashCode() ?: 0)
        result = 31 * result + (tableVariables?.contentHashCode() ?: 0)
        result = 31 * result + (parallelism ?: 0)
        result = 31 * result + (maxBulkSize ?: 0)
        result = 31 * result + (partitionUnit?.hashCode() ?: 0)
        result = 31 * result + (partitionSize ?: 0)
        result = 31 * result + partitioned.hashCode()
        result = 31 * result + (functions?.contentHashCode() ?: 0)
        result = 31 * result + (dependencies?.contentHashCode() ?: 0)
        result = 31 * result + (sources?.contentHashCode() ?: 0)
        result = 31 * result + (targets?.contentHashCode() ?: 0)
        result = 31 * result + (hashKey?.hashCode() ?: 0)
        result = 31 * result + (description?.hashCode() ?: 0)
        result = 31 * result + (deactivatedTs?.hashCode() ?: 0)
        result = 31 * result + (createdTs?.hashCode() ?: 0)
        result = 31 * result + (updatedTs?.hashCode() ?: 0)
        return result
    }

    override fun toString(): String {
        return "TableModel(id=$id, area=$area, vertical=$vertical, name=$name, version=$version, className=$className, transformSqlQuery=$transformSqlQuery, partitionColumns=${partitionColumns?.contentToString()}, tableVariables=${tableVariables?.contentToString()}, parallelism=$parallelism, maxBulkSize=$maxBulkSize, partitionUnit=$partitionUnit, partitionSize=$partitionSize, partitioned=$partitioned, functions=${functions?.contentToString()}, dependencies=${dependencies?.contentToString()}, sources=${sources?.contentToString()}, targets=${targets?.contentToString()}, hashKey=$hashKey, description=$description, deactivatedTs=$deactivatedTs, createdTs=$createdTs, updatedTs=$updatedTs)"
    }


}